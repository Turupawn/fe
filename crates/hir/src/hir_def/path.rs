use super::IdentId;
use crate::{hir_def::Partial, HirDb};

#[salsa::interned]
pub struct PathId<'db> {
    #[return_ref]
    pub segments: Vec<Partial<IdentId<'db>>>,
}

impl<'db> PathId<'db> {
    pub fn last_segment(self, db: &'db dyn HirDb) -> Partial<IdentId> {
        self.segments(db).last().copied().unwrap_or_default()
    }

    pub fn len(self, db: &dyn HirDb) -> usize {
        self.segments(db).len()
    }

    pub fn is_ident(self, db: &dyn HirDb) -> bool {
        self.len(db) == 1
    }

    pub fn self_ty(db: &'db dyn HirDb) -> Self {
        let self_ty = Partial::Present(IdentId::make_self_ty(db));
        Self::new(db, vec![self_ty])
    }

    pub fn from_ident(db: &'db dyn HirDb, ident: IdentId<'db>) -> Self {
        Self::new(db, vec![Partial::Present(ident)])
    }

    pub fn push(self, db: &'db dyn HirDb, segment: IdentId<'db>) -> Self {
        let mut segments = self.segments(db).clone();
        segments.push(Partial::Present(segment));
        Self::new(db, segments)
    }

    pub fn pretty_print(self, db: &dyn HirDb) -> String {
        self.segments(db)
            .iter()
            .map(|seg| {
                seg.to_opt()
                    .map_or("_".to_string(), |id| id.data(db).clone())
            })
            .collect::<Vec<_>>()
            .join("::")
    }
}
