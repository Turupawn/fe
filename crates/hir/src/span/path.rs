use parser::ast;

use super::{define_lazy_span_node, LazySpanAtom};

define_lazy_span_node!(
    LazyPathSpan,
    ast::Path,
    @idx {
        (segment, LazyPathSegmentSpan),
    }
);

define_lazy_span_node!(LazyPathSegmentSpan);
impl<'db> LazyPathSegmentSpan<'db> {
    pub fn into_atom(self) -> LazySpanAtom<'db> {
        LazySpanAtom(self.0)
    }
}
