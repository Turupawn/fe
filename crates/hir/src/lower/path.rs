use parser::ast;

use super::FileLowerCtxt;
use crate::hir_def::{IdentId, Partial, PathId};

impl<'db> PathId<'db> {
    pub(super) fn lower_ast(ctxt: &mut FileLowerCtxt<'db>, ast: ast::Path) -> Self {
        let mut segments = Vec::new();
        let db = ctxt.db();

        for seg in ast.into_iter() {
            let segment = match seg.kind() {
                Some(ast::PathSegmentKind::Ingot(_)) => Some(IdentId::make_ingot(db)),
                Some(ast::PathSegmentKind::Super(_)) => Some(IdentId::make_super(db)),
                Some(ast::PathSegmentKind::SelfTy(_)) => Some(IdentId::make_self_ty(db)),
                Some(ast::PathSegmentKind::Self_(_)) => Some(IdentId::make_self(db)),
                Some(ast::PathSegmentKind::Ident(ident)) => Some(IdentId::lower_token(ctxt, ident)),
                None => None,
            }
            .into();
            segments.push(segment);
        }

        Self::new(db, segments)
    }

    pub(super) fn lower_ast_partial(
        ctxt: &mut FileLowerCtxt<'db>,
        ast: Option<ast::Path>,
    ) -> Partial<Self> {
        ast.map(|ast| Self::lower_ast(ctxt, ast)).into()
    }
}
