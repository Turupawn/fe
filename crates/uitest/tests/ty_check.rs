use std::path::Path;

use dir_test::{dir_test, Fixture};
use driver::DriverDataBase;
use fe_compiler_test_utils::snap_test;

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/fixtures/ty_check",
    glob: "**/*.fe"
)]
fn run_ty_check(fixture: Fixture<&str>) {
    let mut db = DriverDataBase::default();
    let path = Path::new(fixture.path());

    let input_file = db.standalone(path, fixture.content());
    let top_mod = db.top_mod(input_file);

    let diags = db.run_on_top_mod(top_mod);
    let diags = diags.format_diags(&db);
    snap_test!(diags, fixture.path());
}

#[cfg(target_family = "wasm")]
mod wasm {
    use wasm_bindgen_test::wasm_bindgen_test;

    use super::*;

    #[dir_test(
        dir: "$CARGO_MANIFEST_DIR/fixtures/ty_check",
        glob: "*.fe",
        postfix: "wasm"
    )]
    #[dir_test_attr(
        #[wasm_bindgen_test]
    )]
    fn run_ty_check(fixture: Fixture<&str>) {
        let mut db = DriverDataBase::default();
        let path = Path::new(fixture.path());

        let input_file = db.standalone(path, fixture.content());
        let top_mod = db.top_mod(input_file);
        db.run_on_top_mod(top_mod);
    }
}
