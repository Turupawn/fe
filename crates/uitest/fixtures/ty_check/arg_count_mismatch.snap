---
source: crates/uitest/tests/ty_check.rs
expression: diags
input_file: fixtures/ty_check/arg_count_mismatch.fe
---
error[8-0023]: argument number mismatch
  ┌─ arg_count_mismatch.fe:7:8
  │
2 │ fn add(_ x: usize, _ y: usize) -> usize {
  │    --- function defined here
  ·
7 │     add(1)
  │        ^^^ expected 2 arguments, but 1 given

error[8-0023]: argument number mismatch
  ┌─ arg_count_mismatch.fe:9:8
  │
2 │ fn add(_ x: usize, _ y: usize) -> usize {
  │    --- function defined here
  ·
9 │     add(1, 2, 3)
  │        ^^^^^^^^^ expected 2 arguments, but 3 given

error[8-0033]: `add` is not a method
   ┌─ arg_count_mismatch.fe:24:5
   │
17 │     fn add(_ a: usize, _ b: usize) -> usize {
   │        --- function defined here
   ·
24 │     a.add(1)
   │     ^^^^^^^^
   │     │ │
   │     │ `add` is an associated function, not a method
   │     help: use associated function syntax instead: `Adder::add`
   │
   = note: to be used as a method, a function must have a `self` parameter

error[8-0033]: `add` is not a method
   ┌─ arg_count_mismatch.fe:25:5
   │
17 │     fn add(_ a: usize, _ b: usize) -> usize {
   │        --- function defined here
   ·
25 │     a.add(1, 2)
   │     ^^^^^^^^^^^
   │     │ │
   │     │ `add` is an associated function, not a method
   │     help: use associated function syntax instead: `Adder::add`
   │
   = note: to be used as a method, a function must have a `self` parameter
